src/config_parser.ml

```
@@ -152,6 +228,168 @@ let dhcp_range end_of_directive =
+        fail
+          (Fmt.str "Invalid domain name: %s: %a" e
+             Fmt.(list ~sep:(any ".") string)
+             labels)
+  in
+  let dhcp_host_item =
+    choice ~failure_msg:"Bad dhcp-host argument"
+      [
+        id_thing;
+        net_set_thing;
+        tag_thing;
+        mac_addr;
+        ipv4_addr;
+        (* TODO: ipv6_addr ; *)
+        lease_time;
+        ignore_thing;
+        hostname;
+      ]
+  in
+  sep_by1 (char ',') dhcp_host_item <* end_of_directive >>= fun items ->
+  (* Process items:
+     - We can have at most one id thing except and id and id:* whose semantics
+       are very unclear. Thus we should probably forbid that combination.
+     - For net:/set: we can have as many as we like.
+     - We can as well have as many tag: as we like.
+     - There is also no limit on mac addresses.
+     - There can be at most one ipv4 address.
+     - There can be at most one lease time.
+     - The ignore thing will set a flag. Repeating it is redundant if harmless.
+     - There can be at most one domain name.
+     The option parser in dnsmasq will not enforce much of this. Instead, the
+     last value will overwrite previous values if only one value makes sense.
+     We should do better. *)
+  let thing =
+    List.fold_left
+      (fun config item ->
+        match item with
+        | (`Any_client_id | `Client_id _) as id ->
+            if Option.is_some config.id then
+              Log.warn (fun m ->
+                  m "Redundant id in --dhcp-host. Ignoring previous value!");
+            { config with id = Some id }
+        | `Set set -> { config with sets = set :: config.sets }
+        | `Tag tag -> { config with tags = tag :: config.tags }
+        | `Macaddr mac -> { config with macs = mac :: config.macs }
+        | `Ipv4addr ipv4 ->
+            if Option.is_some config.ipv4 then
+              Log.warn (fun m ->
+                  m "Redundant ipv4 in --dhcp-host. Ignoring previous value!");
```

Same, would fail hard.